feat(node): Add more detailed OTEL docs (#10872)
getsentry/sentry-docs: docs/platforms/javascript/common/opentelemetry/custom-setup.mdx

---
title: Using a Custom OpenTelemetry Setup
description: "How to use use a completely custom OpenTelemetry setup with Sentry."
supported:
  - javascript.nextjs
  - javascript.node
  - javascript.aws-lambda
  - javascript.azure-functions
  - javascript.connect
  - javascript.express
  - javascript.fastify
  - javascript.gcp-functions
  - javascript.hapi
  - javascript.koa
  - javascript.nestjs
  - javascript.sveltekit
  - javascript.astro
  - javascript.remix
notSupported:
  - javascript
---

If you already have OpenTelemetry set up yourself, you can also use your existing setup.

<Note>

> This setup is only needed if you need to add custom span processors or exporters to your OpenTelemetry setup. Most Sentry users can follow standard SDK setup to get OpenTelemetry data with their SDK.

Reviewer: ```suggestion
This setup is only needed if you need to add custom span processors or exporters to your OpenTelemetry setup. Most Sentry users can follow a standard SDK setup to get OpenTelemetry data with their SDK.
```

  // The SentrySampler will use this to determine which traces to sample
  tracesSampleRate: 1.0,
});

// Note: This could be BasicTracerProvider or any other provider depending on
// how you are using the OpenTelemetry SDK
const provider = new NodeTracerProvider({
  // Ensure the correct subset of traces is sent to Sentry
  // This also ensures trace propagation works as expected
  sampler: sentryClient ? new SentrySampler(sentryClient) : undefined,
});

// Ensure spans are correctly linked & sent to Sentry
provider.addSpanProcessor(new SentrySpanProcessor());

provider.register({
  // Ensure trace propagation works
  // This relies on the SentrySampler for correct propagation
  propagator: new SentryPropagator(),
  // Ensure context & request isolation are correctly managed
  contextManager: new Sentry.SentryContextManager(),
});

// Validate that the setup is correct
Sentry.validateOpenTelemetrySetup();
```

Make sure that all <PlatformLink to="/opentelemetry/required-instrumentation/">Required OpenTelemetry Instrumentation</PlatformLink> is set up correctly - otherwise, the Sentry SDK may not work as expected!

## Using Sentry for Error Monitoring only

> If you have a custom OpenTelemetry setup and only want to use Sentry for error monitoring, you can skip adding the `SentrySpanProcessor`. You will still need to add the `SentryContextManager`, `SentryPropagator` and `SentrySampler` to your setup even if you do not want to sent any tracing data to Sentry. See below on an explanation of why this is needed.

Reviewer 2: ```suggestion
If you have a custom OpenTelemetry setup and only want to use Sentry for error monitoring, you can skip adding the `SentrySpanProcessor`. You will still need to add the `SentryContextManager`, `SentryPropagator` and `SentrySampler` to your setup even if you don't want to send any tracing data to Sentry. See below on an explanation of why this is needed.
```

## Why do I need to do this?

> In order for the Sentry SDK to work as expected, and for it to be able to be in sync with OpenTelemetry, we need a few components to be in place. These are needed even if you only use Sentry for error monitoring:

Reviewer 2: ```suggestion
In order for the Sentry SDK to work as expected, and for it to be in sync with OpenTelemetry, we need a few components to be in place. These are needed even if you only use Sentry for error monitoring:
```

> - **SentryContextManager**: This is needed to ensure that the OpenTelemetry context is in sync with Sentry.

Reviewer 2: l: feel free to omit but maybe an example helps to reason about this:
```suggestion
- **SentryContextManager**: Ensures that the OpenTelemetry context is in sync with Sentry, for example to correctly isolate data between simultaneous requests.
```

provider.register({
  // Ensure trace propagation works
  // This relies on the SentrySampler for correct propagation
  propagator: new SentryPropagator(),
  // Ensure context & request isolation are correctly managed
  contextManager: new Sentry.SentryContextManager(),
});

// Validate that the setup is correct
Sentry.validateOpenTelemetrySetup();
```

Make sure that all <PlatformLink to="/opentelemetry/required-instrumentation/">Required OpenTelemetry Instrumentation</PlatformLink> is set up correctly - otherwise, the Sentry SDK may not work as expected!

## Using Sentry for Error Monitoring only

If you have a custom OpenTelemetry setup and only want to use Sentry for error monitoring, you can skip adding the `SentrySpanProcessor`. You will still need to add the `SentryContextManager`, `SentryPropagator` and `SentrySampler` to your setup even if you don't want to send any tracing data to Sentry. See below on an explanation of why this is needed.

## Why do I need to do this?

In order for the Sentry SDK to work as expected, and for it to be in sync with OpenTelemetry, we need a few components to be in place. These are needed even if you only use Sentry for error monitoring:

> 

Author: I would like to somehow keep a reference that this is also relevant if you do not use tracing in Sentry - this is repeatedly a point of confusion for users, why they need these things if they only use sentry for error monitoring 🤔

- **SentryContextManager**: This is needed to ensure that the OpenTelemetry context is in sync with Sentry.

> - **SentrySampler**: This is needed to ensure that the Sentry `tracesSampleRate` is respected. Even if you do not use Sentry for tracing, you still need this in order for trace propagation to work as expected. See <PlatformLink to='/opentelemetry/sampler'>Using a Custom OpenTelemetry Sampler</PlatformLink> if you want to use a custom sampler.

Reviewer 2: l:
```suggestion
- **SentrySampler**: Ensures that the Sentry `tracesSampleRate` is respected. Even if you do not use Sentry for tracing, you still need this in order for trace propagation to work as expected. See <PlatformLink to='/opentelemetry/sampler'>Using a Custom OpenTelemetry Sampler</PlatformLink> if you want to use a custom sampler.
```

> - **SentrySampler**: Ensures that the Sentry `tracesSampleRate` is respected. Even if you do not use Sentry for tracing, you still need this in order for trace propagation to work as expected. See <PlatformLink to='/opentelemetry/sampler'>Using a Custom OpenTelemetry Sampler</PlatformLink> if you want to use a custom sampler.

Author: I think it's better to avoid things like "this" as link texts 🤔 maybe we can use something else a bit shorter?

> - **SentryPropagator**: This is needed to ensure that trace propagation works correctly.

Reviewer 2: l:
```suggestion
- **SentryPropagator**: Ensures that trace propagation works correctly.
```

> - <PlatformLink to="/opentelemetry/required-instrumentation">Required Instrumentation</PlatformLink>: These are needed to ensure that trace propagation works correctly.

Reviewer 2: ```suggestion
- <PlatformLink to="/opentelemetry/required-instrumentation">Required Instrumentation</PlatformLink>: Ensures that trace propagation works correctly.
```

<Note>
  Trace propagation is needed for Sentry to automatically connect services (e.g.
  frontend and backend, or different backend services) together. This makes it
  possible to see related errors across services.
  <PlatformLink to="/tracing/trace-propagation">Learn more about Trace Propagation.</PlatformLink>
</Note>

> The following is only needed if you use Sentry for tracing:

Reviewer 2: To confirm, `SentrySpanProcessor` is needed in addition to everything else if Sentry is also used for tracing, right? I'd point this out more clearly:

```suggestion
In addition, the following is only needed if you also use Sentry for tracing:
```

> - **SentrySpanProcessor**: This is needed to ensure spans are correctly sent to Sentry.

Reviewer 2: l:
```suggestion
- **SentrySpanProcessor**: Ensures that spans are correctly sent to Sentry.
```
